### PR TITLE
Allow users to provide a Registry URL for storing images

### PR DESCRIPTION
**Pull request by wtam2018**
This is a move away from the Quay.io registry and switching to using the internal Openshift registry.

./odo pipelines bootstrap --image-repo quay.io/kmcdermo/testing # that uses Quay.io
./odo pipelines boostrap --image-repo docker.io/bigkevmcd/testing # that would use Docker
./odo pipelines bootstrap --image-repo image-registry.openshift-image-registry.svc:5000/project/app  or simply,
./odo pipelines bootstrap --image-repo project/app  
This should use the internal OpenShift registry.

The provided URL should be minimally validated that it has multiple components.

Fixes https://issues.redhat.com/browse/GITOPS-47

### CODE REVIEW

File: pkg/odo/cli/pipelines/bootstrap.go [hunk 1]
```diff
@@ -67,17 +70,32 @@ func (bo *BootstrapOptions) Validate() error {
...
+		GitRepo:                  bo.gitRepo,
+		ImageRepo:                bo.imageRepo,
+		InternalRegistryHostname: bo.internalRegistryHostname,
+		Prefix:                   bo.prefix,
+		QuayAuthFileName:         bo.quayIOAuthFilename,
+		QuayUserName:             bo.quayUsername,
+		SkipChecks:               bo.skipChecks,
+	}
+	if err := checkOptionalRegistryParams(options); err != nil {
+		return err
```
Comment (bigkevmcd) [inline pkg/odo/cli/pipelines/bootstrap.go]: if they're not optional, maybe they shouldn't be in the `BootstrapOptions` type?

File: pkg/pipelines/bootstrap.go [hunk 2]
```diff
@@ -197,3 +220,40 @@ func marshalOutputs(out io.Writer, outputs []interface{}) error {
...
+	// check for spaces
+	for _, v := range components {
+		if strings.TrimSpace(v) == "" {
+			return false, "", fmt.Errorf(errorMsg, o.ImageRepo)
+		}
+	}
+
+	if len(components) == 2 {
+		if components[0] == "docker.io" || components[0] == "quay.io" {
+			// we recconize docker.io and quay.io.  It is missing one component
```
Comment (bigkevmcd) [inline pkg/pipelines/bootstrap.go]: s/recconize/recognize/

File: pkg/pipelines/bootstrap.go [hunk 3]
```diff
@@ -197,3 +220,40 @@ func marshalOutputs(out io.Writer, outputs []interface{}) error {
 	}
 	return nil
 }
+
+// validateImageRepo validates the input image repo.  It determines if it is
+// for internal registry and prepend internal registry hostname if neccessary.
+func validateImageRepo(o *BootstrapOptions) (bool, string, error) {
+	components := strings.Split(o.ImageRepo, "/")
+	errorMsg := "failed to parse image repo:%s, expected image repository in the form <registry>/<username>/<repository> or <project>/<app> for internal registry"
```
Comment (bigkevmcd) [inline pkg/pipelines/bootstrap.go]: I'd probably just extract this to a function that generated the error with a message?

File: pkg/pipelines/bootstrap.go [hunk 4]
```diff
@@ -112,16 +116,35 @@ func Bootstrap(o *BootstrapOptions) error {
 	route := routes.Generate(namespaces["cicd"])
 	outputs = append(outputs, route)
 
-	//  Create Service Account, Role, Role Bindings, and ClusterRole Bindings
-	outputs = append(outputs, createRoleBindings(namespaces)...)
+	// Create Service Account
+	sa := createServiceAccount(meta.NamespacedName(namespaces["cicd"], saName))
+
+	if isInternalRegistry {
```
Comment (bigkevmcd) [inline pkg/pipelines/bootstrap.go]: Could probably move this into a separate function that returned the correct set of outputs?

File: pkg/pipelines/bootstrap.go [hunk 5]
```diff
@@ -64,6 +68,13 @@ func Bootstrap(o *BootstrapOptions) error {
 			return errors.New("failed due to Tekton Pipelines or Triggers are not installed")
 		}
 	}
+
+	// Validate image repo
```
Comment (bigkevmcd) [inline pkg/pipelines/bootstrap.go]: These comments don't add anything to the code?

File: pkg/pipelines/bootstrap.go [hunk 6]
```diff
@@ -197,3 +220,40 @@ func marshalOutputs(out io.Writer, outputs []interface{}) error {
...
+func validateImageRepo(o *BootstrapOptions) (bool, string, error) {
+	components := strings.Split(o.ImageRepo, "/")
+	errorMsg := "failed to parse image repo:%s, expected image repository in the form <registry>/<username>/<repository> or <project>/<app> for internal registry"
+
+	// repo url has minimum of 2 components
+	if len(components) < 2 {
+		return false, "", fmt.Errorf(errorMsg, o.ImageRepo)
+	}
+
+	// check for spaces
```
Comment (bigkevmcd) [inline pkg/pipelines/bootstrap.go]: It's not checking for spaces tho'?

It's ensuring that the paths aren't empty?

File: pkg/odo/cli/pipelines/bootstrap.go [hunk 7]
```diff
@@ -67,17 +70,32 @@ func (bo *BootstrapOptions) Validate() error {
...
 	}
 	return pipelines.Bootstrap(&options)
 }
 
+func checkOptionalRegistryParams(options pipelines.BootstrapOptions) error {
```
Comment (bigkevmcd) [inline pkg/odo/cli/pipelines/bootstrap.go]: Shouldn't this only be checked if we're using Quay?

File: pkg/pipelines/triggers/pipelinerun_test.go [hunk 8]
```diff
@@ -21,7 +21,7 @@ func TestCreateDevCDPipelineRun(t *testing.T) {
 			Resources:          createDevResource("REPLACE_IMAGE:$(params.gitref)"),
 		},
 	}
-	template := createDevCDPipelineRun(sName)
+	template := createDevCDPipelineRun(sName, "REPLACE_IMAGE")
```
Comment (bigkevmcd) [inline pkg/pipelines/triggers/pipelinerun_test.go]: This is probably not an indicative value for the parameter?

how about...

```suggestion
	template := createDevCDPipelineRun(sName, "example.com:5000/testing/testing")
```

File: pkg/odo/cli/pipelines/bootstrap.go (hunk 1 above)
Comment (wtam2018) [reply pkg/odo/cli/pipelines/bootstrap.go]: Options in BootstrapOptions can be required or optionals.    I don't think BootstrapOptions implies all of the fields are optionals.   We could rename it to BootstrapParameters ....

File: pkg/pipelines/bootstrap.go (hunk 3 above)
Comment (wtam2018) [reply pkg/pipelines/bootstrap.go]: This function validates and retrieves image repo from the options.   We can change the function name to "getImageRepo"   I am not quite sure I am following what you are saying.   Please clarify.

File: pkg/pipelines/bootstrap.go (hunk 2 above)
Comment (wtam2018) [reply pkg/pipelines/bootstrap.go]: Fixed

File: pkg/pipelines/bootstrap.go (hunk 4 above)
Comment (wtam2018) [reply pkg/pipelines/bootstrap.go]: Moved

File: pkg/pipelines/bootstrap.go (hunk 5 above)
Comment (wtam2018) [reply pkg/pipelines/bootstrap.go]: Removed

File: pkg/pipelines/bootstrap.go (hunk 6 above)
Comment (wtam2018) [reply pkg/pipelines/bootstrap.go]: Fixed comment.   Add new check for white spaces.  And test cases.

File: pkg/odo/cli/pipelines/bootstrap.go (hunk 7 above)
Comment (wtam2018) [reply pkg/odo/cli/pipelines/bootstrap.go]: Actually, quay user name is not needed anymore (now that we have --image-repo).    dockerconfigjson is checked after we parse image-repo and determine whether it is an internal or external registry.   We validate dockerconfigjson filename only if it is needed (using external registry).

File: pkg/pipelines/triggers/pipelinerun_test.go (hunk 8 above)
Comment (wtam2018) [reply pkg/pipelines/triggers/pipelinerun_test.go]: Fixed.

File: pkg/odo/cli/pipelines/bootstrap.go (hunk 1 above)
Comment (bigkevmcd) [reply pkg/odo/cli/pipelines/bootstrap.go]: Renaming it sounds good, certainly when I conceived of the idea, I intended for the main driver function to accept required parameters as args, and optional things as an options type argument.

We've shifted away from this, but options still remains, which feels odd.

File: pkg/pipelines/bootstrap.go (hunk 3 above)
Comment (bigkevmcd) [reply pkg/pipelines/bootstrap.go]: Sorry, the errorMsg could just be 

```go
func validationError(imageRepo string) error {
  return fmt.Errorf("failed to parse image repo: %s, expected image repository in the form <registry>/<username>/<repository> or <project>/<app> for internal registry", imageRepo)
}
```

File: pkg/pipelines/bootstrap.go [hunk 9]
```diff
@@ -132,6 +143,43 @@ func createRoleBindings(ns map[string]string) []interface{} {
 	return out
 }
 
+// createManifestsForImageRepo creates manifects like namespaces, secret, and role bindng for using image repo
```
Comment (bigkevmcd) [inline pkg/pipelines/bootstrap.go]:
```suggestion
// createManifestsForImageRepo creates manifests like namespaces, secret, and role bindings for using image repo
```

File: pkg/pipelines/bootstrap.go [hunk 10]
```diff
@@ -112,16 +114,25 @@ func Bootstrap(o *BootstrapOptions) error {
 	route := routes.Generate(namespaces["cicd"])
 	outputs = append(outputs, route)
 
-	//  Create Service Account, Role, Role Bindings, and ClusterRole Bindings
-	outputs = append(outputs, createRoleBindings(namespaces)...)
+	// Create Service Account
+	sa := createServiceAccount(meta.NamespacedName(namespaces["cicd"], saName))
+
+	// Create decret, role binding, namespaces for using image repo
```
Comment (bigkevmcd) [inline pkg/pipelines/bootstrap.go]:
```suggestion
	// Create secret, role binding, namespaces for using image repo
```

File: pkg/pipelines/bootstrap.go [hunk 11]
```diff
@@ -112,16 +114,25 @@ func Bootstrap(o *BootstrapOptions) error {
 	route := routes.Generate(namespaces["cicd"])
 	outputs = append(outputs, route)
 
-	//  Create Service Account, Role, Role Bindings, and ClusterRole Bindings
-	outputs = append(outputs, createRoleBindings(namespaces)...)
+	// Create Service Account
```
Comment (bigkevmcd) [inline pkg/pipelines/bootstrap.go]: We can drop these comments, it just repeats what the code itself says.

File: pkg/odo/cli/pipelines/bootstrap.go (hunk 1 above)
Comment (wtam2018) [reply pkg/odo/cli/pipelines/bootstrap.go]: Renamed.

File: pkg/pipelines/bootstrap.go (hunk 3 above)
Comment (wtam2018) [reply pkg/pipelines/bootstrap.go]: Fixed.

**Comment by wtam2018** [comment]
@bigkevmcd I'll take care of the conflicts.

File: pkg/pipelines/bootstrap.go (hunk 10 above)
Comment (wtam2018) [reply pkg/pipelines/bootstrap.go]: Fixed

File: pkg/pipelines/bootstrap.go (hunk 11 above)
Comment (wtam2018) [reply pkg/pipelines/bootstrap.go]: Fixed

**Pull request merged**